### PR TITLE
[REVIEW]Add function to create hashed vocabulary file from raw vocabulary

### PR DESCRIPTION
**Pull request by VibhuJawa**
This pr closes part of https://github.com/rapidsai/cudf/issues/5799  by upstreaming the [`perfect_hash.py`](https://github.com/rapidsai/clx/blob/267c6d30805c9dcbf80840f222bf31c5c4b7068a/python/clx/analytics/perfect_hash.py) to `cudf`. 

Please note I don't understand the details of the inner workings of `perfect_hash.py` and this is more of a one to one port of the file with minimal code changes. 

To ensure correctness i ensured that we get the same result as `perfect-hash.py`  ( [vocab_hash.txt](https://github.com/rapidsai/cudf/blob/910e5276e2a7b734652d05b18e9fbf9b5571fa25/python/cudf/cudf/tests/data/vocab_hash/ground_truth_vocab_hash.txt)) created on the vocabulary [`bert-base-uncased-vocab.txt`]( python/cudf/cudf/tests/data/vocab_hash/bert-base-uncased-vocab.txt) 

The main change here is that I have gotten rid of the `non-compact` code path as that caused failures like at [issue](https://github.com/rapidsai/cudf/issues/5760#issuecomment-664671013) . 

### TODO: 
- [x] Add function
- [x] Add Test to ensure equivalence 
- [x] Add ChangeLog  

### Previous Problems:
Below have been addressed now by sampling nonspecial symbols. 
1.  Adding this test will : 
a. Add `30s` to the test suite
b. Add `1.8 Mb` because of the `ground truth` and `vocabulary files` 

We can reduce both if the above are unexpectable by sampling the vocabulary to lesser words. 

### Updated  PR:
Below have been addressed now by sampling nonspecial symbols. 
1.  Adding this test will : 
a. Add `1.5 s` to the test suite
b. Add `112 kb` because of the `ground truth` and `vocabulary files`

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by VibhuJawa** [comment]
This pr should be ready for an initial review.

**Comment by kkraus14** [comment]
> 1. Adding this test will :
>    a. Add `30s` to the test suite
>    b. Add `1.8 Mb` because of the `ground truth` and `vocabulary files`
> 
> We can reduce both if the above are unexpectable by sampling the vocabulary to lesser words.

The 30s testing addition is un-ideal but acceptable, but the 1.8 Mb is too large for a git repo. Could we use a much smaller file for testing (which would presumably speed up the testing anyway)?

**Comment by VibhuJawa** [comment]
> > 1. Adding this test will :
> >    a. Add `30s` to the test suite
> >    b. Add `1.8 Mb` because of the `ground truth` and `vocabulary files`
> > 
> > We can reduce both if the above are unexpectable by sampling the vocabulary to lesser words.
> 
> The 30s testing addition is un-ideal but acceptable, but the 1.8 Mb is too large for a git repo. Could we use a much smaller file for testing (which would presumably speed up the testing anyway)?

Sure, I will do a random of the bert-vocabulary to draw 5% of the data, I think it should cover all edge cases. Will update the PR.

**Comment by VibhuJawa** [comment]
Switched to 5% sampling of nonspecial symbols, special symbols will have to be kept IMO.  Anyways, we should be good now as it only adds 112kb  and 1.5 s to testing.

**Review by kkraus14 — APPROVED** [review]

File: python/cudf/cudf/core/column/string.py [hunk 1]
```diff
@@ -4201,6 +4201,8 @@ def subword_tokenize(
         ----------
         hash_file : str
             Path to hash file containing vocabulary of words with token-ids.
+            This can be created from the raw vocabulary
+            using the `cudf.utils.hash_vocab_utils.hash_vocab` function
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py:4205]: Need to double tick in docstrings for proper sphinx formatting:
```suggestion
            using the ``cudf.utils.hash_vocab_utils.hash_vocab`` function
```

**Comment by kkraus14** [comment]
CI currently blocked by https://github.com/dask/distributed/issues/4177

**Pull request merged**